evoting/lib/election.go

```diff
@@ -181,54 +186,77 @@ func (e *Election) Box() (*Box, error) {
 }
 
 // Mixes returns all mixes created by the roster conodes.
-func (e *Election) Mixes() ([]*Mix, error) {
-	client := skipchain.NewClient()
+func (e *Election) Mixes(s *skipchain.Service) ([]*Mix, error) {
 
-	block, err := client.GetSingleBlockByIndex(e.Roster, e.ID, 0)
+	// Traversing forward one by one might be expensive in a large
+	// election. It's better to search for the Mix transactions
+	// from the end of the skipchain
+	block, err := s.GetDB().GetLatest(s.GetDB().GetByID(e.ID))
 	if err != nil {
 		return nil, err
 	}
 
 	mixes := make([]*Mix, 0)
 	for {
 		transaction := UnmarshalTransaction(block.Data)
-		if transaction != nil && transaction.Mix != nil {
+		if transaction != nil && transaction.Mix == nil && transaction.Partial == nil {
+			// we're done
+			break
+		}
+
+		if transaction.Mix != nil {
+			// append to the mixes array
 			mixes = append(mixes, transaction.Mix)
 		}
 
-		if len(block.ForwardLink) <= 0 {
-			break
+		// keep iterating back
+		block, err = s.GetSingleBlock(&skipchain.GetSingleBlock{ID: block.BackLinkIDs[0]})
```

Comment:
Why don't you use `s.GetDB().GetByID` here?

Reply:
Fixed